service ca: Remove controller-specific configuration

marun opened the pull request:
Controllers managed by the service ca operator are intended to inherit the operator's `GenericControllerConfig` rather than supporting controller-specific configuration.

marun commented:
/hold

marun commented:
Rebased

marun commented:
Rebased

stlaz commented:
/lgtm

marun commented:
/hold cancel

marun commented:
@sttts Can this merge?

sttts commented on servicecertsigner/v1alpha1/types.go:
```diff
@@ -89,19 +73,12 @@ type ServiceCertSignerOperatorConfigSpec struct {
 	// 2. this config
 	// +kubebuilder:pruning:PreserveUnknownFields
-	ConfigMapCABundleInjectorConfig runtime.RawExtension `json:"configMapCABundleInjectorConfig"`
+	CABundleInjectorConfig runtime.RawExtension `json:"caBundleInjectorConfig"`
```
what does this change mean? Isn't this breaking an existing API?

marun replied:
If this breaks an existing API, but the API is alpha so wouldn't that be ok?

However, maybe it makes more sense to remove this API entirely: 
 - The only thing this config is used for today is disabling authn and authz the the controller (because it doesn't expose endpoints) and setting the path of the ca bundle.
 - What is the value of supporting user overrides of these options given that they are effectively details that should be internal to the operator?
    - The lack of endpoints is an implementation detail
    - The deployment of the injector controller is entirely controlled by the operator, so overriding the bundle path from that set in the deployment would break the controller.

I'm also not clear why the bundle injecting controller shouldn't just watch the signing secret, which would remove the need for the operator to restart the controller whenever the signing secret changes. If the injection controller should not have access to the signing secret for security reasons, then the ca bundle it requires could be written to a configmap instead of being written to the signing secret. I didn't make this change in the current PR series because it seemed tangential to the goal of supporting injection for webhook configs, but I don't think it would be that complicated a change.

deads2k commented:
Create a new value for the new type and keep the old in place.  We don't document how to adjust this anywhere.

I will ask a separate question: do we actually need to expose these options anywhere?  Why would a user change it?

marun commented:
> Create a new value for the new type and keep the old in place. We don't document how to adjust this anywhere.

Understood.  So the existing types will just linger indefinitely?

> I will ask a separate question: do we actually need to expose these options anywhere? Why would a user change it?

That was my thought in a previous [comment](https://github.com/openshift/api/pull/506#discussion_r344313846). But I started to question my assumption that the controller didn't need configuration - was `GenericControllerConfig` a  mandatory requirement for operator-managed controllers? If not, I'm happy to configure the injection controllers without a configuration type and one can be added in the future if users demand it.

marun commented:
- Updated to add a new type and leave the old ones in place
- Removed `CABundleFile` field - this will be hard-coded or supplied by an argument to the command rathe than provided via config since it is entirely dependent on pod configuration.

stlaz commented:
@deads2k @marun since this is alpha, will be we at least able to remove the API in later versions, say, 4.4/4.5? If so, perhaps we should add that to the deprecation note and proceed with the removal in the later version mentioned.

sttts commented on servicecertsigner/v1alpha1/types.go:
```diff
@@ -89,13 +70,24 @@ type ServiceCertSignerOperatorConfigSpec struct {
 	// +kubebuilder:pruning:PreserveUnknownFields
 	APIServiceCABundleInjectorConfig runtime.RawExtension `json:"apiServiceCABundleInjectorConfig"`
 
+	// configMapCABundleInjectorConfig has been deprecated in favor of caBundleInjectorConfig.
```
isn't there some Deprecated syntax we usually use in APIs?

marun replied:
`DEPRECATED[: <reason]` appears to be the most common form.

sttts commented on servicecertsigner/v1alpha1/types.go:
```diff
@@ -89,13 +70,24 @@ type ServiceCertSignerOperatorConfigSpec struct {
+	// 1. hardcoded default
+	// 2. this config
+	// +kubebuilder:pruning:PreserveUnknownFields
+	CABundleInjectorConfig runtime.RawExtension `json:"caBundleInjectorConfig"`
```
What are these rawextensions?

marun commented:
I've revised this PR to add deprecation comments to the controller-specific configuration types and fields, and will be updating the operator to reuse the operator controller configuration for the controllers that it manages.

sttts commented on servicecertsigner/v1alpha1/types.go:
```diff
@@ -11,6 +11,7 @@ import (
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // ServiceServingCertSignerConfig provides information to configure a serving serving cert signing controller
+// DEPRECATED: will be removed soon
```
use xyz instead

marun commented:
@sttts @stlaz I propose that the types and fields marked by deprecation warnings instead be removed immediately. You can confirm the assertions below by reviewing the [proposed removal of their use](https://github.com/openshift/service-ca-operator/pull/88).

- The types are alpha and internal-only - they do not provide a way for a customer to influence controller behavior:
  - For each of the 3 controllers it manages, the operator creates a configmap that includes a serialized instance of the controller-specific configuration type (e.g. `ServiceServingCertSignerConfig`)
  - The controller-specific config is sourced [only from bindata](https://github.com/openshift/service-ca-operator/blob/master/pkg/operator/sync_common.go#L204)
  - Each controller mounts the bindata-sourced configmap and deserializes the contained conttroller-specific configuration type
- The fields are unused. There is no support for reading controller-specific configuration types from the operator configuration.

sttts commented:
/lgtm
/approve

sttts commented:
/retest

sttts commented:
/lgtm
